### PR TITLE
Implement signup

### PR DESCRIPTION
**Pull request by Charitie**
Backend signup implementation

### CODE REVIEW

File: lib/helpers/validateSignupData.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import Schema from 'validate'
+
+const validateSignupData = data => {
+  const userData = new Schema({
```
Comment (Charitie) [inline lib/helpers/validateSignupData.js]: you can take the `Schema` definition outside the function. this will fix the codeClimate issue.

File: app/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import uuid from 'uuid/v4'
+import db from '../../database/models'
+import validateSignupData from '../../lib/helpers/validateSignupData'
+import encryptPassword from '../../lib/helpers/encrypt'
+
+class UserController {
+  static async signup(req, res) {
+    const errors = validateSignupData(req.body)
+    // eslint-disable-next-line no-console
+    console.log(errors)
```
Comment (Rodgers-M) [inline app/controllers/userController.js]: Kindly remove the `console.log` statement. as well as the comment above it.

File: lib/helpers/validateSignupData.js (hunk 1 above)
Comment (Rodgers-M) [inline lib/helpers/validateSignupData.js]: You can take the `Schema` definition out of the function. this will fix the CodeClimate issue.

File: app/controllers/userController.js (hunk 2 above)
Comment (Charitie) [reply app/controllers/userController.js]: Okay

File: lib/helpers/encrypt.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import bcrypt from 'bcrypt'
+
+class EncryptData {
+  static generateHash(password) {
+    return bcrypt.hashSync(password, bcrypt.genSaltSync(10), null)
```
Comment (brianmugweru) [inline lib/helpers/encrypt.js:5]: Set up a variable for 10 at the top and make it a constant

File: lib/helpers/encrypt.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import bcrypt from 'bcrypt'
+
+class EncryptData {
```
Comment (brianmugweru) [inline lib/helpers/encrypt.js:3]: I believe since this is oop, the encryptData will always expect some data to either hash or compare... I mean, this could always define a constructor that sets property data in the case unhashed password that should be hashed within the flow or compared to... Would it make sense to have unhashedPassword as a property @Rodgers-M

File: app/routes/index.js [hunk 5]
```diff
@@ -1,10 +1,14 @@
...
+import userRoutes from './user'
 
 const routes = app => {
   // pass the routes to the app for use
   // e.g app.use('/events', eventsRoutes);
 
+  app.use(userRoutes)
+
+
   app.get('/users', async (req, res) => {
```
Comment (brianmugweru) [inline app/routes/index.js]: Not part of this pr, but lets move it to a users file in routes i.e routes/users.js

File: app/routes/index.js [hunk 6]
```diff
@@ -1,10 +1,14 @@
 import HttpStatus from 'http-status'
 import db from '../../database/models'
+import userRoutes from './user'
```
Comment (brianmugweru) [inline app/routes/index.js:2]: Lets change name from userRoutes to authRoutes

File: app/routes/user.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import express from 'express';
+
+import UserController from '../controllers/userController';
+
+const Router = express.Router();
+
+Router.post('/signup', UserController.signup)
```
Comment (brianmugweru) [inline app/routes/user.js]: Instead of doing validation in the controllers, lets have a validation middleware that calls next if validation passes

File: app/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    if (errors.length) {
+      return res.status(400).json({ message: errors[0].message })
+    }
+
+    // eslint-disable-next-line camelcase
+    const { first_name, last_name, email, password } = req.body
+    const hashedPassword = encryptPassword.generateHash(password)
+    const userData = {
```
Comment (brianmugweru) [inline app/controllers/userController.js]: Lets make this into a transformer.js file or function

File: app/controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import uuid from 'uuid/v4'
+import db from '../../database/models'
+import validateSignupData from '../../lib/helpers/validateSignupData'
```
Comment (brianmugweru) [inline app/controllers/userController.js]: Lets make this a middleware

File: app/routes/user.js (hunk 7 above)
Comment (Rodgers-M) [reply app/routes/user.js]: Makes sense. But what advantage does it give over having it as a helper function? 

I do understand that "ideally"  it's not the work of the controller to handle the business logic, like the `validation` in this case... But our current structure doesn't have `services` which should handle the business logic.

File: app/routes/index.js (hunk 5 above)
Comment (Rodgers-M) [reply app/routes/index.js]: sure we will. it was just a dummy route for testing the Models implementation.

File: lib/helpers/encrypt.js (hunk 4 above)
Comment (Rodgers-M) [reply lib/helpers/encrypt.js:3]: It would make sense having a constructor if we intended to instantiate the class, which we don't anyway coz all of its methods are static.

> Would it make sense to have unhashedPassword as a property

I think it won't because the value is not used anywhere else in the class outside the method and so having it as an argument to the method makes more sense.

**Comment by brianmugweru** [comment]
> Backend signup implementation

Commit messages should be in present form and should touch on what has been updated in a file... You have update and updated as commit messages...

File: app/routes/user.js (hunk 7 above)
Comment (brianmugweru) [reply app/routes/user.js]: > I do understand that "ideally" it's not the work of the controller

This is why, every dev looking at it would expect validations to be part of a middleware... and not having the controller doing so much work leads to the code being more cleaner and readable.

> But our current structure doesn't have services which should handle the business logic

Structure should not affect how validation should be done

File: app/controllers/userController.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      password: hashedPassword,
+    }
+    try {
+      const user = await db.User.findOne({ where: { email } })
+
+      if (user) {
+        return res.status(409).json({ message: 'email already exists' })
+      }
+      await db.User.create(userData)
+      return res.json({ message: 'a user attempted to signup to the app' })
```
Comment (dessHub) [inline app/controllers/userController.js]: Is this a success message? It seems more of a failure message.

File: .vscode/settings.json [hunk 11]
```diff
@@ -0,0 +1,2 @@
+{
```
Comment (brianmugweru) [inline .vscode/settings.json]: Add this to .gitignore file

File: app/controllers/userController.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import UserService from '../services/userService'
```
Comment (brianmugweru) [inline app/controllers/userController.js:1]: This is actually way better than what we had before, I like the services idea more...

File: app/controllers/userController.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+
+class UserController {
+  static async signup(req, res) {
+    return UserService.createUser(req, res)
```
Comment (brianmugweru) [inline app/controllers/userController.js]: Let's only pass req.body to createUser function... And then pass the results of createUser down to .then function and handle errors with catch e.g.
```
await UserService.createUser(req.body)
  .then(transformToRemovePasswordProperty)
  .then(user => res.status(200).send({ data: user })
  .catch(next)
```
- I hope we already have a documented way in which we are getting our req.body user data so that the data conforms to a specific design

File: app/routes/index.js (hunk 6 above)
Comment (brianmugweru) [reply app/routes/index.js:2]: ?

File: app/services/userService.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import uuid from 'uuid/v4'
+import db from '../../database/models'
+import encryptPassword from '../../lib/helpers/encrypt'
+
+class UserService {
+  static async createUser(req, res) {
```
Comment (brianmugweru) [inline app/services/userService.js]: Lets change this to only recieve param user
``` createUser(user) ```

Comment (brianmugweru) [reply app/services/userService.js]: Errors thrown should be passed on to next() in the controller layer

Comment (brianmugweru) [reply app/services/userService.js]: Anything returned should be picked up by the .then function in controller level above

Comment (brianmugweru) [reply app/services/userService.js]:
```
const userData = addUuid(user) => Object.assign({},
{
  id: uuid(),
  ...userData
})
```
``` 
const hashPassword = userData => 
  Object.assign({ 
    ...userData, 
    password: encryptPassword.generateHash(password) 
  })
```

```
const ensureUserDoesNotExist => findUserFunction => userBody => {
  if(await findUserFunction({ where: { email: userBody.email }}) === null) throw ({ message: 'user already exists' })
  return userBody
}
```
```
const createUser = async userData =>
  await db.User.create(userData)
```

- All this are methods that could be used to clean up the piece of code below, make it more readable and better. Since the way forward is OOP, we could use method chaining to achieve that combined with a bit of functional programming... The endgoal is to leave the codebase better before moving on

File: lib/middlewares/validateSignupData.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    required: true,
+    length: { min: 6 },
+  },
+}
+
+const validateSignupData = (req, res, next) => {
+  const userData = new Schema(signUpData)
+  const errors = userData.validate(req.body)
+  if (errors.length) {
+    return res.status(400).json({ message: errors[0].message })
```
Comment (brianmugweru) [inline lib/middlewares/validateSignupData.js:33]: We can have a uniform error hanlder for all the errors, so this allows us to pass errors to function next()
e.g ```errors.length ? next({ message: errors[0].message }) : next()```

**Review by brianmugweru — CHANGES_REQUESTED** [review]

**Comment by brianmugweru** [comment]
Lets add a pr description with steps on how to manual test

**Comment by brianmugweru** [comment]
At least some e2e tests won't do any harm

File: app/services/userService.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    {
```
Comment (brianmugweru) [inline app/services/userService.js:8]: Just realized, we've got curly braces for function createUser... Do they have a purpose?

**Review by brianmugweru — CHANGES_REQUESTED** [review]

File: app/routes/index.js (hunk 6 above)
Comment (Rodgers-M) [reply app/routes/index.js:2]: renaming the routes to `auth` isn't a bad idea, but that means we only limit the routes to authentication only.  Just a thought, will the application at any given point need the functionality to fetch user(s)? let's say for example get the users who have responded to a particular event?

File: app/controllers/userController.js (hunk 13 above)
Comment (Rodgers-M) [reply app/controllers/userController.js]: I think I get what you mean here. 

_you might need to revisit `async/await` though 😄 here is a  [refresher](https://javascript.info/async-await#error-handling)_ 

am thinking of a `CustomError` class that extends the built-in `Error ` that can accept both a `message` and a `status code`. so this class will be used to throw errors with an explicit status code as well as the message. This will enable the global error handler defined to return detailed error messages as well as the correct status code instead of the default 500.

so we could have something like this 
```
// userService 
const user = db.User.findOne({where: { email }})
if(user){
   // Throw an error here. it will be handled by the try/catch block in the controller and passed to next
   throw new CustomError(409, 'User with given email already exists')
}

// userController
static async createUser(req, res, next){
  try{
  await UserService.createUser(req.body)
  return res.status(201).json({message: 'signup success'})
}catch(error)
// in the case the UserService.createUser method threw an error or something went wrong then it will be caught here .
next(error)
}
```

**Review by brianmugweru — APPROVED** [review]
